core/psh/sysexec.c

```diff
@@ -25,6 +27,63 @@ void psh_sysexecinfo(void)
 }
 
 
+static int psh_sysexec_wldargmatch(char *wld, int argc, char **argv, char wldend, int argmaxlen)
+{
+	int len = 0, ii;
+	char *currc = wld, *nextc, *cmdend;
+
+	cmdend = strchr(wld, wldend);
```

Comment:
Yes you are right. It would be much purer form of defining where to end arguments comparison

However i chose the _end sign_ approach because of its universality for both file read and macro read:
- commands from file should have new line character at the end (fgets reads them) so there will always be token when valid command is read. Why not use it?
- commands from macro variable are in one line. Isn`t precalculating the command end then checking if it is reached in `psh_sysexec_wldargmatch` code doubling? Either way there will be check in mentioned function.

Reply:
> Yes you are right. It would be much purer form of defining where to end arguments comparison
> 
> However i chose the _end sign_ approach because of its universality for both file read and macro read:
> 
> * commands from file should have new line character at the end (fgets reads them) so there will always be token when valid command is read. Why not use it?

not true if the last line will not have newline at the end
(it wasn't listed as a requirement and You shouldn't depend on it. Actually, there should probably be test for files ending with "\n" and not ending with "\n"). It would probably be better to strip "\n" before calling this function (this function should be called with the same "wildcard" representation regardless of it's origin).

> * commands from macro variable are in one line. Isn't precalculating the command end then checking if it is reached in `psh_sysexec_wldargmatch` code doubling? Either way there will be check in mentioned function.

You have already computed the end in case of `PSH_WILDCARDS` parsing (by `strchr` in `while` condition), so calling `strchr` here is repeating Yourself and decreases performance.

The usual way for these kinds of parsing is to replace the token character with `\0` so the callee can use standard string functions (like `strlen`) but I assume we don't want to change the `PSH_WILDCARDS` to keep them in `.rodata` - that's why I haven't proposed it before.

Reply:
Yes, you are right. My approach is flawed.

I will rewrite it so that the end of command will be passed as pointer to last character of command, not by expected value of last character.